[dnf5] AssertionError: Add information about assertion source location (macros)

jrohel opened the pull request.

jrohel committed:
AssertionError: Add information about assertion source location (macros)

It is nice to see where the assertion came from.
Now name of file, source line and function/method name are passed
with the assertion.

`std::source_location` has been supported since C++20 and is not yet
implemented in Clang. So the only solution is to use __FILE__, __LINE__,
__PRETTY_FUNCTION__/__func__.

Functions that desire to obtain this information about the call site
must use macros so that predefined macros like __LINE__ are expanded
in the context of the caller.

jrohel committed:
Rewrite specialized assert functions into macros

Needed to obtain the place of origin of the assertion.
Renamed because macros do not use namespaces.

libdnf::assert_same_base -> macro libdnf_assert_same_base
libdnf::rpm::solv::GoalPrivate::assert_resolved -> libdnf_assert_goal_resolved

jrohel committed:
rpm::Transaction::Impl: Use macro to assert that transaction item is set

lukash commented on include/libdnf/common/exception.hpp:
```diff
@@ -20,42 +20,63 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+         : throw libdnf::AssertionError(      \
+               #condition, LIBDNF_LOCATION, message))
+
+#define LIBDNF_ASSERTION_MACROS 1
```
Why are you adding two versions of the macros? To get rid of `fmt::format()` in public headers? What are the reasons to do that? You are adding complexity and confusion because of that and there's no explanation for it in the docstrings or anywhere else...

There are more places where it is desirable to use formatting in headers and the number of them will grow with the codebase so that's the direction we should be heading unless there's an issue with it.

Besides, if we change the standard to C++20 and use its formatting library, this will be a moot point?

jrohel replied:
>Why are you adding two versions of the macros? To get rid of fmt::format() in public headers?

Yes, I want to get rid of `fmt::format` in public headers. I can try to write better explanation. And better naming of macros?

>Besides, if we change the standard to C++20 and use its formatting library, this will be a moot point?

Yes. I've been thinking about a standard text formatting library in C++20. But it's not implemented in gcc or clang these days. We have to wait. In addition, libdnf is now only compatible with C++17.
I'm preparing a PR https://github.com/rpm-software-management/libdnf/pull/1362 with code modifications for compatibility with C++20 and we'll see.

By the way, `format` is more strict in C++20. Requires the format string to be a constant expression.

jrohel replied:
OK. I removed macros without `fmt::format`. Thus `fmt::format` remains in the public header.

jrohel committed:
Remove PROJECT_SOURCE_DIR prefix from __FILE__ macro used in asserts

Prefer location-independent relative paths instead of exposing directory
structure of the build machine.
Uses the preprocessor option `-fmacro-prefix-map=OLD=NEW`.
It is available since GCC 8 and Clang 10.

lukash commented on include/libdnf/common/exception.hpp:
```diff
@@ -24,38 +24,63 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+/// @exception libdnf::AssertionError Thrown always.
+#define libdnf_throw_assertion(msg_format, ...) \
+    (throw libdnf::AssertionError(              \
+        nullptr, LIBDNF_LOCATION, fmt::format(msg_format, ##__VA_ARGS__)))
```
As the assertion error is not meant to be programmatically processed (e.g. caught and printed in a custom way), I don't think it's necessary to store all the string separate in the message and override `what()` to put them together. Instead you could do it directly here, either by simply letting the compiler to concat string literals if we limit ourselves to `msg_format` also being a literal:
```C++
#define libdnf_throw_assertion(msg_format, ...) \
    (throw libdnf::AssertionError(              \
        fmt::format(__FILE__ ":" __LINE__ ": " __PRETTY_FUNCTION__ ": Assertion failed: " msg_format, ##__VA_ARGS__)))
```

Or via joining `std::string`s:
```C++
#define libdnf_throw_assertion(msg_format, ...) \
    (throw libdnf::AssertionError(              \
        fmt::format(std::string(__FILE__ ":" __LINE__ ": " __PRETTY_FUNCTION__ ": Assertion failed: ") + msg_format, ##__VA_ARGS__)))
```

This would cut down on the number of LoCs and complexity, to me it seems a simple approach is better as the complex one is just not going to be used anyway.

jrohel replied:
@lukash
1. `__LINE__` is not a string literal, but a number. I can do macro magic.
E.g.:
```
#define libdnf_stringify(x) #x
#define libdnf_tostring(x) libdnf_stringify(x)
__FILE__ ":" libdnf_tostring(__LINE__) ":"
```
Do you really want it?

2. The `__PRETTY_FUNCTION__` extension is not a string literal. It is implemented similarly to the standard `__func__`. As C++ standard says `__func__` is a static const functional-local predefined variable. I can't convert to a string literal.

3. Even if `__PRETTY_FUNCTION__` was a string literal, I don't want to concatenate assert message at compile time. Why fill binary files with pre-prepared messages? To make it bigger? We don't need to optimize string concatenations for speed. Assert occurs only rarely, ideally never.

lukash replied:
Ok, sorry, I haven't realized. It's still a possibility to concat the final string in the constructor though. Given the amount of fragments to the whole thing, running them through the format call seems cleaner (this one again assumes `msg_format` is a string literal):
```C++
#define libdnf_throw_assertion(msg_format, ...) \
    (throw libdnf::AssertionError(              \
        fmt::format("{}:{}: {}: Assertion failed: " msg_format, __FILE__, __LINE__, __PRETTY_FUNCTION__, ##__VA_ARGS__)))
```

j-mracek commented on dnfdaemon-server/transaction.cpp:
```diff
@@ -20,7 +20,7 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 #include <fmt/format.h>
-
+#include <stdexcept>
 
```
Please keep 2 spaces around namespace

jrohel replied:
Fixed

j-mracek commented on dnfdaemon-server/transaction.cpp:
```diff
@@ -44,9 +44,9 @@ RpmTransactionItemActions transaction_package_to_action(const libdnf::base::Tran
+            throw std::logic_error(fmt::format("Unexpected action in RpmTransactionItem: {}", tspkg.get_action()));
     }
-    throw libdnf::AssertionError("Unknown action in RpmTransactionItem: {}", tspkg.get_action());
+    throw std::logic_error(fmt::format("Unknown action in RpmTransactionItem: {}", tspkg.get_action()));
```
Is the change required because `libdnf::AssertionError` is not a part of public API?

jrohel replied:
Yes. That was the original intention.
But now the assert macros are public. So they can be used/abused in other programs - microdnf, dnf-daemon. So I can change this code and use asserts from libdnf.

jrohel replied:
`libdnf_throw_assertion` is now used.

lukash replied:
To be honest making these macros a part of our public API is so far what I dislike the most about this.

j-mracek commented:
LGTM
